src/qibolab/_core/instruments/keysight/qcs.py

```diff
@@ -0,0 +1,172 @@
+"""Qibolab driver for Keysight QCS instrument set."""
+
+from collections import defaultdict
+from functools import reduce
+from typing import ClassVar, Optional
+
+import numpy as np
+from keysight import qcs
+
+from qibolab._core.components import AcquisitionChannel, Config, DcChannel, IqChannel
+from qibolab._core.execution_parameters import AveragingMode, ExecutionParameters
+from qibolab._core.identifier import ChannelId, Result
+from qibolab._core.instruments.abstract import Controller
+from qibolab._core.pulses import PulseId
+from qibolab._core.sequence import InputOps, PulseSequence
+from qibolab._core.sweeper import ParallelSweepers
+from qibolab._core.unrolling import Bounds
+
+from .pulse import (
+    process_acquisition_channel_pulses,
+    process_dc_channel_pulses,
+    process_iq_channel_pulses,
+)
+from .results import fetch_result, parse_result
+from .sweep import process_sweepers
+
+NS_TO_S = 1e-9
+
+BOUNDS = Bounds(
+    waveforms=1,
+    readout=1,
+    instructions=1,
+)
+
+__all__ = ["KeysightQCS"]
+
+
+class KeysightQCS(Controller):
+    """Driver for interacting with QCS controller server."""
+
+    bounds: str = "qcs/bounds"
+
+    # Map of QCS virtual channels to QCS physical channels
+    qcs_channel_map: qcs.ChannelMapper
+    # Map of Qibolab channel IDs to QCS virtual channels
+    virtual_channel_map: dict[ChannelId, qcs.Channels]
+    # Map of QCS virtual acquisition channels to QCS state classifiers
+    classifier_map: Optional[dict[qcs.Channels, qcs.MinimumDistanceClassifier]] = {}
+    sampling_rate: ClassVar[float] = (
+        qcs.SAMPLE_RATES[qcs.InstrumentEnum.M5300AWG] * NS_TO_S
+    )
+
+    def connect(self):
+        self.backend = qcs.HclBackend(self.qcs_channel_map, hw_demod=True)
+        self.backend.is_system_ready()
+
+    def create_program(
+        self,
+        sequence: PulseSequence,
+        configs: dict[str, Config],
+        sweepers: list[ParallelSweepers],
+        num_shots: int,
+    ) -> tuple[qcs.Program, list[tuple[int, int]]]:
+        program = qcs.Program()
+
+        # SWEEPER MANAGEMENT
+        probe_channel_ids = {
+            chan.probe
+            for chan in self.channels.values()
+            if isinstance(chan, AcquisitionChannel)
+        }
+
+        (
+            hardware_sweepers,
+            software_sweepers,
+            sweeper_channel_map,
+            sweeper_pulse_map,
+        ) = process_sweepers(sweepers, probe_channel_ids)
+        # Here we are telling the program to run hardware sweepers first, then software sweepers
+        # It is essential that we match the original sweeper order to the modified sweeper order
+        # to reconcile the results at the end
+        sweep = lambda program, sweepers: program.sweep(*sweepers)
+        program = reduce(
+            sweep,
+            software_sweepers,
+            reduce(sweep, hardware_sweepers, program).n_shots(num_shots),
```

Comment:
I would probably drop the initial `program = qcs.Program()` instantiation (first line of the function) and inline it here
```suggestion
            reduce(sweep, hardware_sweepers, qcs.Program()).n_shots(num_shots),
```
to avoid shadowing, particularly because the keyword `program` appears in several different occasions in these few lines (including the `lambda` argument).

Reply:
Implemented in https://github.com/qiboteam/qibolab/pull/944/commits/8c279d1bb6a5c69225d62edaa63138a9b130e3ea